lib/route/nh_encap_seg6_local.c

```diff
@@ -0,0 +1,1202 @@
+	 * disjoined, this allow us to release acquired resources by optional
+	 * attributes and by required attributes independently from each other
+	 * without any interference.
+	 * In other terms, we are sure that we do not release some the acquired
+	 * resources twice.
+	 *
+	 * Note that if an attribute is configured both as required and as
+	 * optional, it means that the user has messed something up in the
+	 * seg6_action_table. Therefore, this check is required for SRv6
+	 * behaviors to work properly.
+	 */
+	invalid_attrs = desc->attrs & desc->optattrs;
+	if (invalid_attrs) {
+		NL_DBG(1, "An attribute cannot be both required AND optional");
+		return -NLE_INVAL;
+	}
+
+	/* parse the required attributes */
+	for (i = SEG6_LOCAL_SRH; i < SEG6_LOCAL_MAX + 1; i++) {
+		if (desc->attrs & SEG6_F_ATTR(i)) {
+			if (!attrs[i])
+				return -NLE_INVAL;
+
+			param = &seg6_action_params[i];
+
+			err = param->parse(attrs, slwt);
+			if (err < 0)
+				goto parse_attrs_err;
+		}
+	}
+
+	/* parse the optional attributes, if any */
+	err = parse_nla_optional_attrs(attrs, slwt);
+	if (err < 0)
+		goto parse_attrs_err;
+
+	return 0;
+
+parse_attrs_err:
+	/* release any resource that may have been acquired during the i-1
+	 * parse() operations.
+	 */
+	destroy_attrs(desc->attrs, i, slwt);
+
+	return err;
+}
+
+static void seg6_local_encap_destructor(void *priv)
+{
+	struct seg6_local_lwt *slwt = priv;
+	unsigned long attrs = slwt->desc->attrs | slwt->parsed_optattrs;
+
+	destroy_attrs(attrs, SEG6_LOCAL_MAX + 1, slwt);
+}
+
+/**
+ * Copied from Linux 6.4: seg6_local_build_state:net/ipv6/seg6_local.c
+ * Author: David Lebrun <david.lebrun@uclouvain.be>
+ */
+static int seg6_local_encap_parse_msg(struct nlattr *nla, struct rtnl_nexthop *nh)
+{
+	struct nlattr *tb[SEG6_LOCAL_MAX + 1];
+	struct rtnl_nh_encap *	rtnh_encap;
+	struct seg6_local_lwt *slwt;
+	int err;
+
+	err = nla_parse_nested(tb, SEG6_LOCAL_MAX, nla, seg6_local_encap_policy);
+	if (err < 0)
+		return err;
+
+	if (!tb[SEG6_LOCAL_ACTION])
+		return -NLE_INVAL;
+
+	rtnh_encap = calloc(1, sizeof(*rtnh_encap));
+	if (!rtnh_encap)
+		return -NLE_NOMEM;
+
+	slwt = calloc(1, sizeof(*slwt));
+	if (slwt == NULL) {
+		free(rtnh_encap);
+		return -NLE_NOMEM;
+	}
+
+	slwt->action = nla_get_u32(tb[SEG6_LOCAL_ACTION]);
+
+	err = parse_nla_action(tb, slwt);
+	if (err < 0)
+		goto err;
+
+	rtnh_encap->priv = slwt;
+	rtnh_encap->ops = &seg6_local_encap_ops;
+
+	nh_set_encap(nh, rtnh_encap);
+
+	return 0;
+
+err:
+	free(slwt);
+	free(rtnh_encap);
+	return err;
+}
+
+static int seg6_local_encap_build_msg(struct nl_msg *msg, void *priv)
+{
+	struct seg6_local_lwt *slwt = priv;
+	struct seg6_action_param *param;
+	unsigned long attrs;
+	int i, err;
+
+	NLA_PUT_U32(msg, SEG6_LOCAL_ACTION, slwt->action);
+
+	attrs = slwt->desc->attrs | slwt->parsed_optattrs;
+
+	for (i = SEG6_LOCAL_SRH; i < SEG6_LOCAL_MAX + 1; i++) {
+		if (attrs & SEG6_F_ATTR(i)) {
+			param = &seg6_action_params[i];
+			err = param->put(msg, slwt);
```

Comment:
put constructs a netlink message to send to kernel, right?

E.g. there is `rtnl_route_nh_get_encap_seg6_local_table()`, but there is no setter. So you cannot create a libnl object with the local-table set, and thus you cannot set the SEG6_LOCAL_TABLE attribute (and the put code is unusable).

Or how this usable?

Reply:
I do not implement setter for SEG6_LOCAL_TABLE - thought it is OK, because not all attributes has setters (rtnl_route_nh_get_encap_mpls_ttl for example)
Do you think we need implement it now?